### ISSUE TITLE
Informer could loss some event

### ISSUE THREAD

**Issue by cy-zheng** [kubernetes/client-go#571]
I found out that informer could loss some short event. For example, when I run a k8s job with a simple command "ls -l", informer receive event list such as PodPending, PodPending, PodPending, PodSucceed. The PodRunning state is lost. 

Is that a bug? Or the informer doesn't guarantee that every event will be received correctly?

**Comment by hzxuzhonghu** [comment]
It is not a bug, its not guaranteed.

**Comment by HeGaoYuan** [comment]
@liggitt @dims @deads2k @smarterclayton @sttts @nikhita @pohly @alexzielenski 

Other k8s experts are also expected to answer this question. Sorry if bother you. Thanks!

**Comment by pohly** [comment]
Informers are guaranteed to report the current state of an object, eventually. They are not guaranteed to report every in-between state. This is a feature, not a bug: Kubernetes is level-driven, not edge-driven, so what matters is the current state, not what happened to get there.

/close

**Comment by HeGaoYuan** [comment]
> Informers are guaranteed to report the current state of an object, eventually. They are not guaranteed to report every in-between state. This is a feature, not a bug: Kubernetes is level-driven, not edge-driven, so what matters is the current state, not what happened to get there.
> 
> /close

Thanks for your reply!

Maybe both of us said is right?

My understand of  level-driven and edge-driven is following. The difference between level-driven and edge-driven is that when informer receive a event from api-server, the level-driven will send object itself to informer, but edge-driven will send variety of object to informer. But both of level-driven and edge-driven could not guarantee that "The PodRunning state event will send". For k8s, what matters is the current state, so could not guarantee "The PodRunning state event will send" does not matter. What matters is the current state is also the reason k8s adopt level-driven. So what @pohly and @hzxuzhonghu said is right.

What I said "Informer will guarantee not miss any event, except one case" is also right? My best evidence actually from @sttts and @mhausenblas 's book Chapter 3 Informers and Caching

![截屏2023-10-11 15 03 22](https://github.com/kubernetes/client-go/assets/16939557/7c9b0161-bd43-4c54-a83d-0d230afd61b2)

So both of us said is right? It is just because k8s api-server will not guarantee to send the PodRunning event, so informer will not receive this event.

**Comment by sttts** [comment]
It's actually a property on the informer side. It does not guarantee that every event coming through a watch request from the apiserver is actually visible to the informer event handlers. The described error handling is such a situation where the lossless watch retry does not succeed and the best the informer can do is to get a latest state via list. Technically you could use watch requests directly and see every event, but it would be brittle in case of (apiserver) failures.

**Comment by HeGaoYuan** [comment]
@sttts 
Thanks for your reply!

Your book is amazing, it helps me a lot! 

**But in your book you said Informer is "edge-triggered, level-driven with additional resync" seems different with what @pohly said "Kubernetes is level-driven, not edge-driven".** Actually I understand what you describe in your book, I think you describe about the behavior of Informer clearly. Informer is triggered when there is a event/change(edge-driven triggers) not a regular checked (level-driven triggers). But when informer is triggered the informer to get the latest of object from cache(cache and api-server/etcd should be thought same), this is level-driven. And there is additional resync period for Informer/Controller between the in-memory cache and the business logic: the registered event handlers will be called for all objects each time this period has passed, it makes Informer/Controller more robust especially reconcile about external system. But there are also many people said "Kubernetes is level-driven, not edge-driven", as @pohly said and the link (https://hackernoon.com/level-triggering-and-reconciliation-in-kubernetes-1f17fe30333d) you give in your book. Maybe they talk about the behavior of api-server? Wish your reply to clear my doubts. 

Wish your explain/reply more specific more better. Forgive me for being such a curious baby

---

<img width="675" alt="截屏2023-10-11 17 28 18" src="https://github.com/kubernetes/client-go/assets/16939557/29d6041a-5197-43fe-a7e7-f3a140d950ee">
n following picture. Informer/Controller is 


![截屏2023-10-11 17 24 24](https://github.com/kubernetes/client-go/assets/16939557/02c97fc3-453f-45ab-a35b-d2f39dd01595)

---

**Comment by sttts** [comment]
> Kubernetes is level-driven, not edge-driven

This is a simplification. The business logic is level-based, that's the important bit. Edges trigger, and resync triggers.

**Comment by HeGaoYuan** [comment]
@sttts  

Thanks again for your reply.

Or cloud please answer my following question? I think the following question's answer will clear all my doubts. Please forgive my long-winded.

If a Pod's real change path is PodPending, PodRuning, PodSucceed. Then whether following is guaranteed?

a. Will the api-server be guaranteed to send to the watch client 3 events with correct object and correct order?
b. Will client-go's Informer's Reflector's watchHandler function (it means not get object from cache) be guaranteed to receive 3 events with correct object and correct order?

https://github.com/kubernetes/client-go/blob/cf4002bbbf855062bb00291127a16c9677a4c643/tools/cache/reflector.go#L696-L708

I think b is right. Please tell me if I am wrong.

If a is right, then I think this issue is because most client will use informer's cache to get the object, maybe client receive 3 event, but maybe the client process very slowly, at 2th event, the cache has changed to PodSucceed, so the client will see PodSucceed, so the client thought it miss the PodRunning.

If a is wrong, then the api-server maybe just send 3 events with PodPending, PodSucceed, PodSucceed. So the client see what api-server just send.

**Comment by sttts** [comment]
a is right, in the happy case because the events are coming directly from etcd. But watches stop and network fails etc.
For b I would have to dive into the code too, don't know by heart where from the watch request to the informer event handlers event can go missing.

**Comment by HeGaoYuan** [comment]
@sttts 

Thanks! Looking forward you dive into the code, then further take a communication with you.

Maybe I can share why I am so persist in this question.

My question seems stupid and meaningless for Deployment. We all know that the Deployment Controller only concern the final state of Pod, so there are not big problems if I miss the PodRuning or any other middle states.

But now because k8s is so popular, there are many operator like https://github.com/GoogleCloudPlatform/spark-on-k8s-operator, https://github.com/kubeflow/training-operator, their CRD is for a short time task job. Most of the job will concern the Pod's state to determine the job's state. If we miss a middle state, it may cause some problem. And there are some development requirements from myself Product Manager, for example, we need to record the Job's creating path to get a better product experience, my direct implementation relies on the middle state.

I thought the api-server will guarantee the middle state and send it. Like the "a. expression" I post above (a. Will the api-server be guaranteed to send to the watch client 3 events with correct object and correct order?). This is the precondition of the following.

Then what we concern is in client-go. There is a classical picture about client-go.

<img width="1047" alt="截屏2023-10-12 11 31 16" src="https://github.com/kubernetes/client-go/assets/16939557/a9fb2fcf-fa38-4922-871c-8d664deccf48">

For this picture, I thought 1), 2), 3), 4) will all guarantee the middle state. 

Because the 7), 8), 9) is in another goroutine, then for 9) when it get the object for key it may loss the middle state.

So for some feature need to concern the middle state, my implementation is to write a synchronous 6), a custome event handlers to process the object from 3), so I think the middle state is guaranteed.

But when I see this issue, the main Member give a short conclusion "It is not a bug, its not guaranteed.". No offense, from the thumbs down numbers, I think this short answer confuse many normal developers. Normal developers maybe wrong, but will you could give a more specific answer. @hzxuzhonghu 

So, that's all  why I am so persist in this question.

**Comment by hzxuzhonghu** [comment]
First, in etcd, it just stores the latest object, not intermediate objects.

let me draft some cases:

Case 1: there are two kube-apiservers, and one operator communicates with a different apiserver from the system components like kube-scheduler, kubelet which mutate a pod status.

So if the operator connected kubeapiserver crash in some time, while at the same time the pod get mutated(maybe many state transitions), so once the operator reconnect to the new restarted kube-apiserver, it definitely can only get the latest pod state, so we say some event missing.


Case 2:  in kube-apiserver, there is a watch cache, IIRC, by default the capacity is 1000,. if the cache is full, the oldest unconsumed event will be evicted. 

So when the operator/client does not consume quickly, it could miss events. And it can also happen in the case that your client disconnects with the kube-apiserver for a long enough time.

**Comment by HeGaoYuan** [comment]
@hzxuzhonghu 

Thanks! Your reply helps me a lot!

Then I think the philosophy of Kubernetes is "what matters is the current state, not what happened to get there". The important is final state/ final consistence.

Maybe I should consider carefully when developing a Operator to try my best to keep this design philosophy.

**Comment by hzxuzhonghu** [comment]
I am a little curious about your use case now

**Comment by HeGaoYuan** [comment]
> I am a little curious about your use case now

@hzxuzhonghu  Like I said here, https://github.com/kubernetes/client-go/issues/571#issuecomment-1758860457.

My use case is like https://github.com/GoogleCloudPlatform/spark-on-k8s-operator, https://github.com/kubeflow/training-operator, to create a CRD about a AI/Machine Learning Training Job. And as a commercial product, we may record the process of Job's LifeCycle to our users. So we need to record when the job is created, when the job is running, when the job is failed, etcd. And the job's status also come from the pod's status, if a the controller only get podpending and podfailed, with no podrunning, the controller will make job's status to jobCreated, JobFailed, without JobRunning, it may confuse our users why my job does not running but failed.

And we need to record the failed reason of the Job, it also comes form the Pod's status. We need to take special action for Pod's evicted situation. But I found that when a pod got evicted, sometimes we can only get that through the pod.Status.Conditions(PodReady is false and reason is Evict), it is a short middle state, then the most strange thing is pod.Status.Conditions will become normal(PodReady became true), at last the Pod would reach failed but the reason sometime not Evicted. So by this observation, the only way I check whether a Pod is Evicted is check the middle state, if I miss it, I will not know the reason why the Pod is failed.

That's first case that comes to my mind. Maybe I can give more general and good case in the future when the CRD is about the short-time Job.

**Comment by HeGaoYuan** [comment]
Maybe I can give more cases for Job CRD.

A node maybe removing by users suddenly. Then the Pod running on the node should be deleted then the final state is Pod not exists. My product manager require that for this case we should report the Job failed but not try to create a new Pod in another Node. Then my direct implementation is to check the deletionTime filed of Pod. If I miss this middle state, I maybe just get the Pod not exists, maybe my program will create a new Pod.

If I following the middle state is not reliable guideline. I should do more thing, for example, when I found I need create a new Pod(compare the current pod with desired pod), in the before I just create a new Pod without any other concern. Now I think I should check the Job's status, if it is running, in this time maybe I should not just Create Pod.

All I want to say is that for a Job CRD, there are many cases,  the easiest and direct implementation/thinking for some feature/requirement  is replying on a middle state. If the middle state is not reliable, for developing a robust system, I should think very carefully to implement it.

**Comment by HeGaoYuan** [comment]
@hzxuzhonghu @sttts @pohly 

I can share more thoughts as a normal developer. It is also another reason why I care about this question so much.

When I started to learn k8s's client-go, I am a little confused about the Resync function in Informer. Actually not just me, I found that at least in the Chinese Internet, there are so many people argue the function of Resync in Informer. Some people thought that the Resync is used for doing a new relist from api-server to get the latest and correct data per Resync time. Some people thought that the Resync is used for doing a new reconcile, where the change/event comes from all the cached data in Informer. I think the second opinion is right.  The book Programming Kubernetes also think second is right. Though second opinion  is right, people will doubt whether the api-server and client-go can guarantee that they will not miss any event. If they missed any event, there will be problem. For example if they missed the Pod's last failed event, the Informer's cache will store PodRunning as the latest's status of Pod, but it is wrong, it will influence all Controller even that only care about the final consistence. That's the reason I think the statement of a and b here are right. https://github.com/kubernetes/client-go/issues/571#issuecomment-1757525498
